File: engine/src/components/Camera.cpp
//// Camera.cpp ///////////////////////////////////////////////////////////////
//
//  zzzzz       zzz  zzzzzzzzzzzzz    zzzz      zzzz       zzzzzz  zzzzz
//  zzzzzzz     zzz  zzzz                    zzzz       zzzz           zzzz
//  zzz   zzz   zzz  zzzzzzzzzzzzz         zzzz        zzzz             zzz
//  zzz    zzz  zzz  z                  zzzz  zzzz      zzzz           zzzz
//  zzz         zzz  zzzzzzzzzzzzz    zzzz       zzz      zzzzzzz  zzzzz
//
//  Author:      Mehdy MORVAN
//  Date:        13/05/2025
//  Description: Source file for the camera component
//
///////////////////////////////////////////////////////////////////////////////

#include "Camera.hpp"

namespace nexo::components {
    [[nodiscard]] glm::mat4 CameraComponent::getProjectionMatrix() const
    {
        if (type == CameraType::PERSPECTIVE) {
            return glm::perspective(glm::radians(fov), static_cast<float>(width) / static_cast<float>(height), nearPlane, farPlane);
        }
        return glm::ortho(0.0f, static_cast<float>(width), static_cast<float>(height), 0.0f, nearPlane, farPlane);
    }

    [[nodiscard]] glm::mat4 CameraComponent::getViewMatrix(const TransformComponent &transf) const
    {
        const glm::vec3 forward = transf.quat * glm::vec3(0.0f, 0.0f, -1.0f);
        const glm::vec3 upVec = transf.quat * glm::vec3(0.0f, 1.0f, 0.0f);
        return glm::lookAt(transf.pos, transf.pos + forward, upVec);
    }

    void CameraComponent::resize(const unsigned int newWidth, const unsigned int newHeight)
    {
        width = newWidth;
        height = newHeight;
        resizing = true;
        if (m_renderTarget) {
            m_renderTarget->resize(newWidth, newHeight);
        }
        pipeline.resize(newWidth, newHeight);
    }


    void CameraComponent::restore(const CameraComponent::Memento& memento)
    {
        if (width != memento.width || height != memento.height) {
            width = memento.width;
            height = memento.height;
            resize(width, height);
        }
        viewportLocked = memento.viewportLocked;
        fov = memento.fov;
        nearPlane = memento.nearPlane;
        farPlane = memento.farPlane;
        type = memento.type;
        clearColor = memento.clearColor;
        main = memento.main;
        m_renderTarget = memento.renderTarget;
    }

    [[nodiscard]] CameraComponent::Memento CameraComponent::save() const
    {
        return {
            width,
            height,
            viewportLocked,
            fov,
            nearPlane,
            farPlane,
            type,
            clearColor,
            main,
            m_renderTarget
        };
    }
}

File: engine/src/scripting/managed/Components/Camera.cs
//// Camera.cs ////////////////////////////////////////////////////////////////
//
//  zzzzz       zzz  zzzzzzzzzzzzz    zzzz      zzzz       zzzzzz  zzzzz
//  zzzzzzz     zzz  zzzz                    zzzz       zzzz           zzzz
//  zzz   zzz   zzz  zzzzzzzzzzzzz         zzzz        zzzz             zzz
//  zzz    zzz  zzz  z                  zzzz  zzzz      zzzz           zzzz
//  zzz         zzz  zzzzzzzzzzzzz    zzzz       zzz      zzzzzzz  zzzzz
//
//  Author:      Thomas PARENTEAU
//  Date:        01/06/2025
//  Description: Source file for the Camera component in C#.
//
///////////////////////////////////////////////////////////////////////////////

using System.Numerics;
using System.Runtime.InteropServices;

namespace Nexo.Components
{
    public enum CameraType
    {
        Perspective = 0,
        Orthographic = 1
    }

    [StructLayout(LayoutKind.Sequential)]
    public struct CameraComponent
    {
        public uint width;
        public uint height;
        [MarshalAs(UnmanagedType.I1)] public bool viewportLocked;
        public float fov;
        public float nearPlane;
        public float farPlane;
        public CameraType type;
        public Vector4 clearColor;
        [MarshalAs(UnmanagedType.I1)] public bool active;
        [MarshalAs(UnmanagedType.I1)] public bool render;
        [MarshalAs(UnmanagedType.I1)] public bool main;
        [MarshalAs(UnmanagedType.I1)] public bool resizing;
        // Note: m_renderTarget is excluded (pointer to shared_ptr)
    }

    [StructLayout(LayoutKind.Sequential)]
    public struct PerspectiveCameraController
    {
        public Vector2 lastMousePosition;
        public float mouseSensitivity;
        public float translationSpeed;
        [MarshalAs(UnmanagedType.I1)] public bool wasMouseReleased;
        [MarshalAs(UnmanagedType.I1)] public bool wasActiveLastFrame;
    }

    [StructLayout(LayoutKind.Sequential)]
    public struct PerspectiveCameraTarget
    {
        public Vector2 lastMousePosition;
        public float mouseSensitivity;
        public float distance;
        public uint targetEntity;
    }
}
